### PR TITLE
Add class for BlackoilWellModel constraints

### PR DESCRIPTION
**Pull request by akva2**
More well code splitting.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/wells/BlackoilWellModelConstraints.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+{
+public:
+    BlackoilWellModelConstraints(const BlackoilWellModelGeneric& well)
+        : well_(well)
+    {}
+
+    /// Return true if any well has a THP constraint.
+    bool hasTHPConstraints() const;
+
+    std::pair<Group::InjectionCMode, double>
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelConstraints.hpp:47]: We must include `<utility>` in order to use `std::pair<>` as a return type.

File: opm/simulators/wells/BlackoilWellModelConstraints.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    bool updateGroupIndividualControl(const Group& group,
+                                      const int reportStepIdx,
+                                      std::map<std::pair<std::string,Opm::Phase>,std::string>& switched_inj,
+                                      std::map<std::string, std::string>& switched_prod,
+                                      GroupState& group_state,
+                                      WellState& well_state,
+                                      DeferredLogger& deferred_logger) const;
+
+private:
+    const BlackoilWellModelGeneric& well_;
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelConstraints.hpp:83]: Should probably rename this object to something like `wellModel_`.

File: opm/simulators/wells/BlackoilWellModelConstraints.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+                if (current_rate > 1e-12)
+                    scale = controls.liquid_target / current_rate;
+                return std::make_pair(Group::ProductionCMode::LRAT, scale);
+            }
+        }
+    }
+
+    if (group.has_control(Group::ProductionCMode::CRAT))
+    {
+        OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "CRAT control for production groups not implemented" , deferred_logger);
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelConstraints.cpp:271]: We should include `<stdexcept>` to bring a declaration of `std::runtime_error` into scope.

File: opm/simulators/wells/BlackoilWellModelConstraints.cpp [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+void BlackoilWellModelConstraints::
+actionOnBrokenConstraints(const Group& group,
+                          const Group::ExceedAction& exceed_action,
+                          const Group::ProductionCMode& newControl,
+                          GroupState& group_state,
+                          DeferredLogger& deferred_logger) const
+{
+    const Group::ProductionCMode oldControl = well_.groupState().production_control(group.name());
+
+    std::ostringstream ss;
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelConstraints.cpp:350]: We should include `<sstream>` to bring a declaration of `std::ostringstream` into scope.  We might also consider switching this to using libfmt instead, but that's probably better suited to a follow-up PR.

**Review by bska — COMMENTED** [review]
A couple of missing headers and a somewhat poor data member name, but other than that this looks good to me.

File: opm/simulators/wells/BlackoilWellModelConstraints.hpp (hunk 1 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelConstraints.hpp]: done

File: opm/simulators/wells/BlackoilWellModelConstraints.hpp (hunk 2 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelConstraints.hpp]: done.

File: opm/simulators/wells/BlackoilWellModelConstraints.cpp (hunk 3 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelConstraints.cpp:302]: done.

File: opm/simulators/wells/BlackoilWellModelConstraints.cpp (hunk 4 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelConstraints.cpp:409]: done.

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  I'll merge this into the master branch now.

**Pull request merged**